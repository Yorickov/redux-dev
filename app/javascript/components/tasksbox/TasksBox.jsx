import React from 'react';
import NewTaskForm from './NewTaskForm';
import Tasks from './Tasks';
import Filter from './Filter';

const TasksBox = () => (
  <div className="col-5">
    <NewTaskForm />
    <Filter />
    <Tasks />
  </div>
);

export default TasksBox;
