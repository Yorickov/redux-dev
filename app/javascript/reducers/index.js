import { createReducer, combineReducers } from '@reduxjs/toolkit';
import _ from 'lodash';
import { reducer as formReducer } from 'redux-form';
import * as actions from '../actions';

const tasksFetchingState = createReducer('none', {
  [actions.fetchTasksRequest]: () => 'requested',
  [actions.fetchTasksFailure]: () => 'failed',
  [actions.fetchTasksSuccess]: () => 'finished',
});

const taskRemovingState = createReducer('none', {
  [actions.removeTaskRequest]: () => 'requested',
  [actions.removeTaskFailure]: () => 'failed',
  [actions.removeTaskSuccess]: () => 'finished',
});

const taskAddingState = createReducer('none', {
  [actions.addTaskRequest]: () => 'requested',
  [actions.addTaskFailure]: () => 'failed',
  [actions.addTaskSuccess]: () => 'finished',
});

const tasks = createReducer({ byId: {}, allIds: [], currentFilterName: 'all' }, {
  [actions.fetchTasksSuccess]: (state, { payload }) => (
    {
      ...state,
      byId: _.keyBy(payload.tasks, 'id'),
      allIds: payload.tasks.map((t) => t.id),
    }
  ),
  [actions.addTaskSuccess]: (state, { payload: { task } }) => {
    const { byId, allIds } = state;
    return {
      ...state,
      byId: { ...byId, [task.id]: task },
      allIds: [task.id, ...allIds],
    };
  },
  [actions.removeTaskSuccess]: (state, { payload: { id } }) => {
    const { byId, allIds } = state;
    return {
      ...state,
      byId: _.omit(byId, id),
      allIds: _.without(allIds, id),
    };
  },
  [actions.toggleTaskState]: (state, { payload: { id } }) => {
    const task = state.byId[id];

    const newState = task.state === 'active' ? 'finished' : 'active';
    const updatedTask = { ...task, state: newState };
    return {
      ...state,
      byId: { ...state.byId, [task.id]: updatedTask },
    };
  },
  [actions.setTasksFilter]: (state, { payload: { filterName } }) => (
    { ...state, currentFilterName: filterName }),
});

export default combineReducers({
  taskRemovingState,
  tasksFetchingState,
  taskAddingState,
  tasks,
  form: formReducer,
});
