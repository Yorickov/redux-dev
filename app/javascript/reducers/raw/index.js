import { createReducer, combineReducers } from '@reduxjs/toolkit';

const counter = createReducer(0, {
  INCREMENT: (state) => state + 1,
});

const text = createReducer('', {
  TEXT_UPDATE: (state, action) => action.payload.text,
  TEXT_RESET: () => '',
});

export default combineReducers({
  counter,
  text,
});
