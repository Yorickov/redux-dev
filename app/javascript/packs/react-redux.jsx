import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import reducers from '../reducers';
import { fetchTasks } from '../actions';
import TasksBox from '../components/tasksbox/TasksBox';

const store = configureStore({
  reducer: reducers,
});

store.dispatch(fetchTasks());

document.addEventListener('DOMContentLoaded', () => {
  ReactDOM.render(
    <Provider store={store}>
      <TasksBox />
    </Provider>,
    document.getElementById('react-redux'),
  );
});
