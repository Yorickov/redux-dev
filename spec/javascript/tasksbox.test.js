import nock from 'nock';
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';
import timeout from 'timeout-then';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import reducers from '../../app/javascript/reducers';
import TasksBox from '../../app/javascript/components/tasksbox/TasksBox';

Enzyme.configure({ adapter: new Adapter() });
axios.defaults.adapter = httpAdapter;
nock.disableNetConnect();

const host = 'http://localhost';

it('Store', async () => {
  const store = configureStore({
    reducer: reducers,
  });

  const vdom = (
    <Provider store={store}>
      <TasksBox />
    </Provider>
  );
  const wrapper = mount(vdom);
  expect(wrapper.render()).toMatchSnapshot();

  const newTaskInput = wrapper.find('input[type="text"]');
  const newTaskSubmit = wrapper.find('button[type="submit"]');

  newTaskInput.simulate('change', { target: { value: 'first task' } });
  expect(wrapper.render()).toMatchSnapshot();

  nock(host)
    .post('/tasks', {
      text: 'first task',
      state: 'active',
    })
    .reply(201, { id: 1, state: 'active', text: 'first task' });

  newTaskSubmit.simulate('submit');
  await timeout(100);
  expect(wrapper.render()).toMatchSnapshot();

  newTaskInput.simulate('change', { target: { value: 'another task' } });
  expect(wrapper.render()).toMatchSnapshot();

  nock(host)
    .post('/tasks', {
      text: 'another task',
      state: 'active',
    })
    .reply(201, { id: 2, state: 'active', text: 'another task' });

  newTaskSubmit.simulate('submit');
  await timeout(100);
  expect(wrapper.render()).toMatchSnapshot();

  const link = wrapper.find('[data-test="task-toggle-state"]').first();
  link.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  const activeFilterButton = wrapper.find('[data-test="task-filter-active"]');
  const finishedFilterButton = wrapper.find('[data-test="task-filter-finished"]');

  activeFilterButton.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  finishedFilterButton.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  const allFilterButton = wrapper.find('[data-test="task-filter-all"]');

  allFilterButton.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  const button = wrapper.find('.close');
  nock(host)
    .delete('/tasks/1')
    .reply(204);

  button.last().simulate('click');
  await timeout(100);
  expect(wrapper.render()).toMatchSnapshot();
});
