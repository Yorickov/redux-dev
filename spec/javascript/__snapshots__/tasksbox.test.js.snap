// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store 1`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
</div>
`;

exports[`Store 2`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value="first task"
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
</div>
`;

exports[`Store 3`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              first task
            </a>
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 4`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value="another task"
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              first task
            </a>
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 5`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              another task
            </a>
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              first task
            </a>
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 6`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              another task
            </a>
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              <s>
                first task
              </s>
            </a>
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 7`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-all"
      type="button"
    >
      All Tasks
    </button>
    Active Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              another task
            </a>
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 8`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-all"
      type="button"
    >
      All Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    Finished Tasks
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              <s>
                first task
              </s>
            </a>
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 9`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              another task
            </a>
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              <s>
                first task
              </s>
            </a>
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 10`] = `
<div
  class="col-5"
>
  <form
    action=""
    class="form-inline"
  >
    <div
      class="form-group mx-sm-3"
    >
      <input
        name="text"
        required=""
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-primary btn-sm"
      disabled=""
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    class="mt-3 d-flex justify-content-around"
  >
    All Tasks
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-active"
      type="button"
    >
      Active Tasks
    </button>
    <button
      class="btn btn-link border-0 p-0"
      data-test="task-filter-finished"
      type="button"
    >
      Finished Tasks
    </button>
  </div>
  <div
    class="mt-3"
  >
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex"
        >
          <span
            class="mr-auto"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              another task
            </a>
          </span>
          <button
            class="close"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
